app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -1,9 +1,49 @@
+/**
+ * navigator.locks is a polyfill for the Locks API, which can be removed from
+ * this file after we drop support for older Firefox versions.
+ * navigator.locks is supported in Firefox 102+.
+ */
+import 'navigator.locks';
 import log from 'loglevel';
 import { captureException } from '@sentry/browser';
 import { isEmpty } from 'lodash';
-import { type MetaMaskStateType, MetaMaskStorageStructure } from './base-store';
-import ExtensionStore from './extension-store';
-import ReadOnlyNetworkStore from './read-only-network-store';
+import { KeyringControllerState } from '@metamask/keyring-controller';
+import { MISSING_VAULT_ERROR } from '../../../../shared/constants/errors';
+import { OnboardingControllerState } from '../../controllers/onboarding';
+import { PreferencesControllerState } from '../../controllers/preferences-controller';
+import { AppStateControllerState } from '../../controllers/app-state-controller';
+import { IndexedDBStore } from './indexeddb-store';
+import type {
+  MetaMaskStateType,
+  MetaMaskStorageStructure,
+  BaseStore,
+  MetaData,
+} from './base-store';
+
+export type Backup = {
+  KeyringController?: KeyringControllerState;
+  OnboardingController?: OnboardingControllerState;
+  PreferencesController?: PreferencesControllerState;
+  AppStateController?: AppStateControllerState;
+};
+
+function makeBackup(state?: MetaMaskStateType): Backup {
+  return {
+    KeyringController: state?.KeyringController,
+    OnboardingController: state?.OnboardingController,
+    PreferencesController: state?.PreferencesController,
+    AppStateController: state?.AppStateController,
+  } as Backup;
+}
+
+function hasVault(state?: MetaMaskStateType): state is MetaMaskStateType {
+  const keyringController = state?.KeyringController as
+    | KeyringControllerState
+    | undefined;
+  return Boolean(keyringController?.vault);
```

Comment:
I don't think we can expect the `keyringController` state to always match the type we're casting to here, and we don't it it to. We can avoid the cast by doing something like this:
```
  const keyringController = state?.KeyringController;
  return (
    isObject(keyringController) &&
    hasProperty(keyringController, 'vault') &&
    Boolean(keyringController?.vault)
  )
```
(using `isObject` and `hasProperty` from `@metamask/utils`)